stuff/cron/plagiarism_detector.py

```diff
@@ -0,0 +1,243 @@
+runs copydetect on submissions recieved. 
+
+Finally it pushes the necessary data to the database
+
+'''
+
+import argparse
+import calendar
+import collections
+import datetime
+import json
+import logging
+import operator
+import os
+import shutil
+import sys
+import tempfile
+from typing import (Any, DefaultDict, Dict, Mapping, NamedTuple, Optional,
+                    Sequence, Set, Tuple, Callable, Iterable)
+
+import boto3  # type: ignore
+sys.path.insert(
+    0,
+    os.path.join(
+        os.path.dirname(os.path.dirname(os.path.realpath(__file__))), "."))
+import lib.db  # pylint: disable=wrong-import-position
+import lib.logs  # pylint: disable=wrong-import-position
+from copydetect import CopyDetector  # type: ignore
+from mysql.connector import errorcode
+
+class Results(NamedTuple):
+    contest_id: int
+    score_1: int
+    score_2: int
+    submission_id_1: int
+    submissino_id_2: int
+    contents: str
+
+# SQL Queries
+
+CONTESTS_TO_RUN_PLAGIARISM_ON = """ SELECT c.`contest_id`, c.`alias`, c.`problemset_id`
+                                    FROM `Contests` as c
+                                    WHERE
+                                        c.`check_plagiarism` = 1 AND
+                                        c.`finish_time` < NOW() - INTERVAL 20 MINUTE AND
+                                        c.`contest_id` NOT IN
+                                            (SELECT p.`contest_id` 
+                                            FROM `Plagiarisms` as p);
+                                """
+
+GET_CONTEST_SUBMISSION_IDS= """ SELECT c.contest_id, s.submission_id, s.problemset_id,
+                                s.problem_id, s.verdict, s.guid, s.language
+                                FROM Submissions as s 
+                                INNER JOIN Contests c ON c.problemset_id = s.problemset_id 
+                                WHERE c.contest_id = %s;
+                            """
+
+INSERT_INTO_PLAGIARISMS = """  
+                                INSERT INTO `Plagiarisms`
+                                (`contest_id`, `score_1`, `score_2` , `submission_id_1` , `submission_id_2`, `contents`)
+                                VALUES
+                                (%s, %s, %s, %s, %s, %s)
+                            """
+
+#Constants
+START_RED = "<span class='highlight-red'>"
+START_GREEN = "<span class='highlight-green'>"
+END = "</span>"
+
+SubmissionDownloader = Callable[[str, str], None]
+
+class S3SubmissionDownloader:
+    ''' A SubmissionDownloader that can download files from an S3 bucket.  '''
+
+    def __init__(self, bucket_name: str = 'omegaup-backup') -> None:
+        self._bucket = boto3.client('s3').Bucket(bucket_name)
+    
+    def __call__(self, guid: str, destination_path: str) -> None:
+        self._bucket.download_file(f'omegaup/submissions/{guid[:2]}/{guid[2:]}', destination_path)
```

Comment:
```suggestion
    def __call__(self, guid: str, destination_dir: str) -> None:
        self._bucket.download_file(f'omegaup/submissions/{guid[:2]}/{guid[2:]}', os.path.join(destination_dir, guid))
```

this way we get the full guid as the filename to avoid the awkward mapping shenanigans below.

Reply:
Then I should also change the destination_dir in the `download` functions to just have the dir and not the guid. Also, we would need language so the 
 ``def __call__(self, guid: str, destination_dir: str, language: str) -> None:
        self._bucket.download_file(f'omegaup/submissions/{guid[:2]}/{guid[2:]}', os.path.join(destination_dir, guid, ".", language))``

Reply:
sounds good!